innovation_sweet_spots/analysis/prototyping/public_discourse_analysis/pd_data_collection_utils.py

```diff
@@ -0,0 +1,184 @@
+    no_html = [seg.get_text() for seg in segments]
+    return no_html
+
+
+def get_text_segments(articles, tags):
+    """Extracts segments of article text from Guardian articles in html format.
+
+    Args:
+        articles: A list of dicts containing article content.
+        tags: A list of html tags to extract content from.
+
+    Returns:
+        A nested list of article segments.
+    """
+    article_html = [article["fields"]["body"] for article in articles]
+    clean_segments = [extract_text_from_html(article, tags) for article in article_html]
+    return clean_segments
+
+
+def get_article_text(articles, tags):
+    """Extracts full text of an article from Guardian articles in html format.
+
+    Args:
+        articles: A list of dicts containing article content.
+        tags: A list of tags to extract content from.
+
+    Returns:
+        A list that contains full text of each article.
+    """
+    article_segments = get_text_segments(articles, tags)
+    article_text = [" ".join(segment) for segment in article_segments]
+    return article_text
+
+
+def get_article_metadata(filtered_articles, fields_to_extract, id_field = 'id'):
+    """Extracts useful article fields from raw data returned by the Guardian API.
+
+    Args:
+        filtered_articles: A list of raw articles.
+        fields_to_extract: A list of fields to extract data from. The useful fields 
+        are "webUrl", "webTitle" and "webPublicationDate"
+        id_field: A string that referrs to field containing article ID.
+
+    Returns:
+        A dict mapping article IDs to other useful fields.
+    """
+    metadata_dict = defaultdict(dict)
+    for article in filtered_articles:
+        article_id = article[id_field]
+        for field in fields_to_extract:
+            article_field = article[field]
+            metadata_dict[article_id][field] = article_field
+    return metadata_dict
+
+
+def get_article_text_df(filtered_articles, tags):
+    """Extracts article text, deduplicates articles and organises them by year.
+
+    Args:
+        grouped_articles: A list of raw articles.
+        tags: A list of html tags to extract content from.
+
+    Returns:
+        A pandas dataframe with article text, id and year.
+    """
+    article_text = get_article_text(filtered_articles, tags)
+    article_id = [article["id"] for article in filtered_articles]
+    article_df = pd.DataFrame({"id": article_id, "text": article_text}).drop_duplicates()
+    return article_df
+
+
+def subset_articles(article_text, refine_terms, required_terms, text_field='text'):
+    """Extracts a subset of articles that contain specific terms.
+    
+    It is useful for removing irrelevant articles in instances when a broad general 
+    search term is used during data collection, e.g. `hydrogen`.
+    
+    Args:
+        article_text: A pandas dataframe that contains article text and id.
+        refine_terms: A list of terms that we use to disambiguate the general term.
```

Comment:
Sorry, it is not clear to me from this docstring what `refine_terms` does exactly - is it another set of required terms to further narrow down the final list of articles?